ff/serving/client.py

```diff
@@ -2,101 +2,208 @@
-                                                  0),))
-        self._stub = proto.serving_pb2_grpc.FeatureStub(channel)
-
-    def dataset(self, name, version):
-        return Dataset(self._stub, name, version)
-
-    def features(self, features, entities):
-        req = proto.serving_pb2.FeatureServeRequest()
-        for name, value in entities.items():
-            entity_proto = req.entities.add()
-            entity_proto.name = name
-            entity_proto.value = value
-        for (name, version) in features:
-            feature_id = req.features.add()
-            feature_id.name = name
-            feature_id.version = version
-        resp = self._stub.FeatureServe(req)
-        return [parse_proto_value(val) for val in resp.values]
-
+	def __init__(self, host):
+		channel = grpc.insecure_channel(host,
+										options=(('grpc.enable_http_proxy',
+												  0),))
+		self._stub = proto.serving_pb2_grpc.FeatureStub(channel)
+
+	def dataset(self, name, version):
+		return Dataset(self._stub, name, version)
+
+	def features(self, features, entities):
+		req = proto.serving_pb2.FeatureServeRequest()
+		for name, value in entities.items():
+			entity_proto = req.entities.add()
+			entity_proto.name = name
+			entity_proto.value = value
+		for (name, version) in features:
+			feature_id = req.features.add()
+			feature_id.name = name
+			feature_id.version = version
+		resp = self._stub.FeatureServe(req)
+		return [parse_proto_value(val) for val in resp.values]
+
+
+class Stream:
+	def __init__(self, stub, name, version):
+		req = proto.serving_pb2.TrainingDataRequest()
+		req.id.name = name
+		req.id.version = version
+		self.name = name
+		self.version = version
+		self._stub = stub
+		self._req = req
+		self._iter = stub.TrainingData(req)
+
+	def __iter__(self):
+		return self
+
+	def __next__(self):
+		try:
+			return Row(next(self._iter))
+		except StopIteration:
+			raise
+
+	def restart(self):
+		self._iter = self._stub.TrainingData(self._req)
+
+
+class Repeat():
+	def __init__(self, repeat_num, stream):
+		self.repeat_num = repeat_num
+		self.stream = stream
+
+	def __iter__(self):
+		return self
+
+	def __next__(self):
+		try:
+			next_val = next(self.stream)
+		except StopIteration:
+			self.repeat_num -= 1
+			if self.repeat_num > 0:
+				self.stream.restart()
+				next_val = next(self.stream)
+			else:
+				raise
+
+		return next_val
+
+
+class Shuffle:
+	def __init__(self, buffer_size, stream):
+		self.buffer_size = buffer_size
+		self._shuffled_data_list = []
+		self.stream = stream
+		self.setup_buffer()
+
+	def setup_buffer(self):
+		try:
+			for _ in range(self.buffer_size):
+				self._shuffled_data_list.append(next(self.stream))
+		except StopIteration:
+			if len(self._shuffled_data_list) == 0:
+				raise
+
+	def __iter__(self):
+		return self
+
+	def __next__(self):
+		if len(self._shuffled_data_list) == 0:
+			# If theres nothing left in the buffer itll return None
```

Comment:
This comment isn't accurate, if there's nothing left it'll raise an exception that stops iteration